piet-direct2d/src/d2d.rs

```diff
@@ -75,8 +78,16 @@ unsafe impl Send for D2DDevice {}
 ///
 /// This struct is public only to use for system integration in piet_common and druid-shell. It is not intended
 /// that end-users directly use this struct.
-#[derive(Clone)]
-pub struct DeviceContext(ComPtr<ID2D1DeviceContext>);
+pub struct DeviceContext(
+    ComPtr<ID2D1DeviceContext>,
+    AssociativeCache<
```

Comment:
This doesn't feel like the right place to store the cache. `DeviceContext` is intended to be a thin wrapper over the raw winapi object, and exists for integration. I had in mind the field would be added to D2DRenderTarget. Is there a strong reason why that doesn't work?

The question of mutability and cloning is a hard one. I'm open to the idea of removing Clone here, but it exists for a reason: to at some point allow background threads to create expensive resources such as images.

If we do end up with this being extended, then it should be a struct with named fields rather than a tuple.

Reply:
The reason I put it here was because there are 3 different places where `create_solid_color` is called. If the cache were moved up a level into `D2DRenderContext` then the 2 calls in `D2DTextLayout` would not benefit from the cache. I could also look into refactoring those to call through the render context instead, but I have no idea how involved of a change that would be. I agree putting it here doesn't feel right either. Thoughts?

Reply:
Urgh. This is indeed a sticky question. Colors are already a place where the otherwise fairly clean abstraction boundary between text and drawing breaks down. I've taken another look at that code. Probably the cleanest thing to do is to change `D2DTextLayout::draw` to take a reference to D2DRenderTarget instead of DeviceContext, but that might create other problems. If so, a targeted approach would be to give that method a (mutable) reference to the color cache as well.

If this is too deep water, then I'd happily accept a PR that just makes the change for drawing, and consider the text stuff a followup, as it's definitely messier.